src/streaming/downsampler.hh

```diff
@@ -35,8 +35,15 @@ class Downsampler
     writer_configurations() const;
 
   private:
-    std::function<ByteVector(ConstByteSpan, size_t&, size_t&)> scale_fun_;
-    std::function<void(ByteVector&, ConstByteSpan)> average2_fun_;
+    using ScaleFunT = std::function<
+      ByteVector(ConstByteSpan, size_t&, size_t&, ZarrDownsamplingMethod)>;
+    using Average2FunT =
+      std::function<void(ByteVector&, ConstByteSpan, ZarrDownsamplingMethod)>;
+
+    ZarrDownsamplingMethod method_;
+
+    ScaleFunT scale_fun_;
+    Average2FunT average2_fun_;
```

Comment:
Do you need to store scale_fun_ and average2_fun_ in the class? I figured you'd just have a function that took method_ as a parameter and dispatched inside there.

Reply:
Yeah it's kind of awkward since it's templated. We'd have to have a `switch` statement over the data type in the functions where they're called. Although I guess we could do something similar to what we're doing with `method` and pass in `data_type` as well with a templated kernel function. Ripping out the logic and doing that seems like it would be a pain.